internal/handlers/tigris/tigris.go

```diff
@@ -62,46 +67,77 @@ func New(opts *NewOpts) (handlers.Interface, error) {
 
 	return &Handler{
 		NewOpts: opts,
+		pools:   make(map[AuthParams]*tigrisdb.TigrisDB, 1),
 	}, nil
 }
 
+// Close implements HandlerInterface.
+func (h *Handler) Close() {
+	h.rw.Lock()
+	defer h.rw.Unlock()
+
+	for k, p := range h.pools {
+		p.Driver.Close()
+		delete(h.pools, k)
+	}
+}
+
 // DBPool returns database connection pool for the given client connection.
 //
 // Pool is not closed when ctx is canceled.
-//
-// TODO https://github.com/FerretDB/FerretDB/issues/1789
 func (h *Handler) DBPool(ctx context.Context) (*tigrisdb.TigrisDB, error) {
+	connInfo := conninfo.Get(ctx)
+	username, password := connInfo.Auth()
+
+	// do not log client secret
+
+	// replace authentication info only if it is present in the connection
+	ap := h.AuthParams
+	if username != "" && password != "" {
+		ap.ClientID = username
+		ap.ClientSecret = password
+	}
+
+	// fast path
+
 	h.rw.RLock()
-	db := h.db
+	p := h.pools[ap]
 	h.rw.RUnlock()
 
-	if db != nil {
-		return db, nil
+	if p != nil {
+		h.L.Debug("DBPool: found existing pool", zap.String("username", username))
+		return p, nil
 	}
 
+	// slow path
+
 	h.rw.Lock()
 	defer h.rw.Unlock()
 
+	// a concurrent connection might have created a pool already; check again
+	if p = h.pools[ap]; p != nil {
+		h.L.Debug("DBPool: found existing pool (after acquiring lock)", zap.String("username", username))
+		return p, nil
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
 	cfg := &config.Driver{
-		ClientID:     h.ClientID,
-		ClientSecret: h.ClientSecret,
-		Token:        h.Token,
-		URL:          h.URL,
-		Protocol:     driver.GRPC,
+		URL:          ap.URL,
+		ClientID:     ap.ClientID,
+		ClientSecret: ap.ClientSecret,
 	}
-	db, err := tigrisdb.New(ctx, cfg, h.L)
+	p, err := tigrisdb.New(ctx, cfg, h.L)
 	if err != nil {
+		h.L.Warn("DBPool: authentication failed", zap.String("username", username), zap.Error(err))
```

Comment:
Is lazy error how we want to fail authentication? I guess it's the same pattern for pg so that's probably what we do for now :hugs:

Reply:
Yeah, that's okay for now